presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -781,8 +782,7 @@ public PlanWithProperties visitJoin(JoinNode node, Context context)
             PlanWithProperties right;
 
             boolean isCrossJoin = type == INNER && leftSymbols.isEmpty();
-            boolean joinWithNonScalar = node.getRight().accept(new IsScalarPlanVisitor(), null);
-            if ((distributedJoins && !isCrossJoin && !joinWithNonScalar) || type == FULL || type == RIGHT) {
+            if ((distributedJoins && !isCrossJoin && !isScalarJoin(node)) || type == FULL || type == RIGHT) {
```

Comment:
it looks like the previous variable name was wrong as `IsScalarPlanVisitor` was returning `true` when plan contained scalar subquery.